### PR TITLE
query_string: add support for ser/de of matrix style query strings (no bracketed keys)

### PR DESCRIPTION
**Pull request by DevinR528**
If any other test cases would be good to add, let me know. <strike>If the benchmarks are unnecessary, I can remove them too, I just wanted to make sure I wasn't doing anything terrible in the serialization side (I used String instead of Cow) it actually goes faster.</strike>

> For structs it might make sense to do json string (de)serialization instead.

This would mean (de)serializing the struct to json then serializing the string? I'm still a bit confused on the details of this.

### CODE REVIEW

File: src/query_string.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            map,
+            deser::from_matrix_str::<HashMap<usize, String>>(&smap).unwrap()
+        );
+    }
+
+    #[test]
+    fn serde_slice_vec_fails() {
+        let vec_ser = ser::to_matrix_query_string(&[1, 2, 3, 4]).unwrap();
+        // oddly this fails
+        assert!(ser::to_matrix_query_string(&vec![1, 2, 3, 4]).is_err());
```
Comment (jplatte) [inline src/query_string.rs:84]: This should definitely not fail..

File: src/query_string/decode.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
+use std::borrow::Cow;
+use std::collections::btree_map::{BTreeMap, Entry};
+use std::iter::Iterator;
+use std::slice::Iter;
+use std::str;
+
+use super::deser::MatrixDeserializer;
+use super::error::{Error, Result};
```
Comment (jplatte) [inline src/query_string/decode.rs:8]: Please merge these imports like we do elsewhere (one `std::{}`, one `super::{}` import).

File: src/query_string/decode.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+#[derive(Clone, Debug)]
+pub(crate) enum Level<'a> {
+    Nested(BTreeMap<Cow<'a, str>, Level<'a>>),
```
Comment (jplatte) [inline src/query_string/decode.rs:12]: Without knowing for sure, this variant seems to not be relevant for us (is it not for `foo[bar]=x&foo[baz]=y`?).

File: src/query_string/error.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+use serde::ser;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum Error {
+    Serialization(String),
+    Deserialization(String),
+    InvalidJson(String),
+    IoWritter(String),
```
Comment (jplatte) [inline src/query_string/error.rs:16]:
```suggestion
    IoWriter(String),
```

**Comment by DevinR528** [comment]
Unfortunately I sort of retyped the serialize part and copy/pasted the deserialize part, I find the best way to understand something is to type it out. I have a local crate that I moved pasted into `ruma-serde`. I can start another PR working from an actual fork pasted into `ruma-serde` as a module?

**Comment by jplatte** [comment]
Hm, but most of the code is still the same as in serde_urlencoded / serde_qs, right? Could you make a PR with your changes to your own fork of whichever you used as a base and request my review there? I can then do the importing into ruma-serde (I'll likely also include the full history for authorship info).

**Comment by DevinR528** [comment]
> Hm, but most of the code is still the same as in serde_urlencoded / serde_qs, right?

Yes, it was from serde_qs, `serde-urlencoded` dose all the parsing in a separate crate so it seemed like that would be harder to deal with.

You want to be able to compare against the base (serde_qs) right? I've made a commit for every file hopefully that helps break it up.

**Comment by jplatte** [comment]
> `serde-urlencoded` dose all the parsing in a separate crate so it seemed like that would be harder to deal with.

If this is a crate we can reuse for our needs, that would actually be perfect. Which crate is that, though, I can't see anything obvious in `serde_urlencoded`s dependencies (at least on crates.io).

> You want to be able to compare against the base (serde_qs) right?

Yes.

> I've made a commit for every file hopefully that helps break it up.

Not really, no. Having the actual changes split would be useful (remove nested struct handling, remove explicit indexing syntax, update list syntax), but if you don't already have the changes in separate commits like that, I can try reviewing file by file.

**Comment by DevinR528** [comment]
Let me check out serde_urlencode again, with possibly just changing that crates logic to fit our need, in mind. So I guess don't worry about reviewing it yet, I'll let you know what I find, hopefully about 30 min I'll have an idea.

**Comment by DevinR528** [comment]
Ok so I think I can use `serde_urlencode` I'll start a PR on my own fork and let you know when I make some progress. I think I can swap the `PairSerializer` out for our sequence serializer and a few other minor tweaks on the serialization side. The deserialization may be a bit more complicated we'll see.

**Comment by DevinR528** [comment]
Here is the PR https://github.com/DevinR528/serde_urlencoded/pull/1

**Comment by DevinR528** [comment]
It's ready to go do want me to add you as a contributor so I can request a review?

**Comment by jplatte** [comment]
I can leave a review without push access, thanks.

**Comment by jplatte** [comment]
Closing since this has now been implemented. Thanks a lot @DevinR528!

**Pull request closed without merging**